src/main/java/org/jenkinsci/plugins/workflow/cps/steps/ingroovy/GroovyStepExecution.java

```diff
@@ -0,0 +1,149 @@
+    }
+
+    @Nonnull
+    @Override
+    public CpsStepContext getContext() {
+        return context;
+    }
+
+    /**
+     * Obtains the {@link GroovyStep} that captures parameters given to this step.
+     */
+    public GroovyStep getStep() {
+        return step;
+    }
+
+    @Override
+    public boolean start() throws Exception {
+        CpsStepContext cps = getContext();
+        CpsThread t = CpsThread.current();
+
+        // TODO: make sure this class is actually CPS transformed
+
+        Closure body = InvokerHelper.getMethodPointer(this, "call");
+        if (getStep().getDescriptor().takesImplicitBlockArgument())
+            body = body.curry(new Body());
+
+        execution = cps.newBodyInvoker(t.getGroup().export(body))
+//                .withStartAction(/*... maybe a marker for the future?*/)
+                .withCallback(BodyExecutionCallback.wrap(cps))
+                .start();
+
+        return false;
+    }
+
+    @Override
+    public void stop(@Nonnull Throwable cause) throws Exception {
+        execution.cancel(cause);
+    }
+
+    /**
+     * Persist {@link #step} in a form that's serializable.
+     */
+    private void writeObject(ObjectOutputStream oos) throws IOException {
+        oos.defaultWriteObject();
+        oos.writeObject(UninstantiatedDescribable.from(step));
+    }
+
+    /**
+     * Pair up with {@link #writeObject(ObjectOutputStream)} to restore {@link #step}
+     */
+    private void readObject(ObjectInputStream ois) throws IOException, ClassNotFoundException {
+        ois.defaultReadObject();
+        UninstantiatedDescribable ud = (UninstantiatedDescribable)ois.readObject();
+        try {
+            step = (GroovyStep)ud.instantiate();
+        } catch (Exception e) {
+            throw new IOException("Failed to instantiate "+ud,e);
+        }
+    }
+
+    /**
+     * Attempt to resolve a method call as a step invocation like Pipeline Script.
+     */
+    public Object methodMissing(String name, Object args) throws IOException {
+        return getDelegate().invokeMethod(name,args);
+    }
+
+    /**
+     * Attempt to resolve a property access as a step invocation like Pipeline Script.
+     */
+    public Object propertyMissing(String name) throws IOException {
+        return getDelegate().getProperty(name);
+    }
+
+    /**
+     * Delegate that implements the method/property resolution of Pipeline Script.
+     */
+    private GroovyObject getDelegate() throws IOException {
+        // this needs to happen inside CPS VM to work correctly, hence no caching
+        return new CpsScript() {
+            @Override
+            public Object run() {
+                throw new AssertionError();
+            }
+        };
+    }
+
+    /**
+     * Passed to the user-written 'call' method in Groovy as 'body'.
+     * When invoked, executes the body block passed to the step.
+     */
+    private final class Body extends Closure implements Serializable {
```

Comment:
Replaced with `@PersistIn(PROGRAM)` to call an attention to the fact that this is serialized.